src/phoenix/utilities/attributes.py

```diff
@@ -0,0 +1,278 @@
+import inspect
+import json
+from typing import (
+    Any,
+    DefaultDict,
+    Dict,
+    Iterable,
+    Iterator,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Union,
+    cast,
+)
+
+from openinference.semconv import trace
+from openinference.semconv.trace import DocumentAttributes, SpanAttributes
+from typing_extensions import assert_never
+
+DOCUMENT_METADATA = DocumentAttributes.DOCUMENT_METADATA
+LLM_PROMPT_TEMPLATE_VARIABLES = SpanAttributes.LLM_PROMPT_TEMPLATE_VARIABLES
+METADATA = SpanAttributes.METADATA
+TOOL_PARAMETERS = SpanAttributes.TOOL_PARAMETERS
+
+# attributes interpreted as JSON strings during ingestion
+JSON_STRING_ATTRIBUTES = (
+    DOCUMENT_METADATA,
+    LLM_PROMPT_TEMPLATE_VARIABLES,
+    METADATA,
+    TOOL_PARAMETERS,
+)
+
+SEMANTIC_CONVENTIONS: List[str] = sorted(
+    # e.g. "input.value", "llm.token_count.total", etc.
+    (
+        cast(str, getattr(klass, attr))
+        for name in dir(trace)
+        if name.endswith("Attributes") and inspect.isclass(klass := getattr(trace, name))
+        for attr in dir(klass)
+        if attr.isupper()
+    ),
+    key=len,
+    reverse=True,
+)  # sorted so the longer strings go first
+
+
+def unflatten(
+    key_value_pairs: Iterable[Tuple[str, Any]],
+    *,
+    prefix_exclusions: Sequence[str] = (),
+    separator: str = ".",
+) -> Dict[str, Any]:
+    # `prefix_exclusions` is intended to contain the semantic conventions
+    trie = _build_trie(key_value_pairs, separator=separator, prefix_exclusions=prefix_exclusions)
+    return dict(_walk(trie, separator=separator))
+
+
+def flatten(
+    obj: Union[Mapping[str, Any], Iterable[Any]],
+    *,
+    prefix: str = "",
+    separator: str = ".",
+    recurse_on_sequence: bool = False,
+    json_string_attributes: Optional[Sequence[str]] = None,
+) -> Iterator[Tuple[str, Any]]:
+    if isinstance(obj, Mapping):
+        yield from _flatten_mapping(
+            obj,
+            prefix=prefix,
+            recurse_on_sequence=recurse_on_sequence,
+            json_string_attributes=json_string_attributes,
+            separator=separator,
+        )
+    elif isinstance(obj, Iterable):
+        yield from _flatten_sequence(
+            obj,
+            prefix=prefix,
+            recurse_on_sequence=recurse_on_sequence,
+            json_string_attributes=json_string_attributes,
+            separator=separator,
+        )
+    else:
+        assert_never(obj)
+
+
+def has_mapping(sequence: Iterable[Any]) -> bool:
+    for item in sequence:
+        if isinstance(item, Mapping):
+            return True
+    return False
+
+
+def get_attribute_value(
+    attributes: Optional[Mapping[str, Any]],
+    key: str,
+    separator: str = ".",
```

Comment:
Do we need support for other separators?

Reply:
no. this is just to make it explicit